sdk/internal/testframework/recording.go

```diff
@@ -0,0 +1,416 @@
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/uuid"
+	"github.com/dnaeon/go-vcr/cassette"
+	"github.com/dnaeon/go-vcr/recorder"
+	"gopkg.in/yaml.v2"
+)
+
+type Recording struct {
+	SessionName              string
+	RecordingFile            string
+	VariablesFile            string
+	Mode                     RecordMode
+	variables                map[string]*string `yaml:"variables"`
+	previousSessionVariables map[string]*string `yaml:"variables"`
+	recorder                 *recorder.Recorder
+	src                      rand.Source
+	now                      *time.Time
+	Sanitizer                *RecordingSanitizer
+	Matcher                  *RequestMatcher
+	c                        TestContext
+}
+
+const (
+	alphanumericBytes           = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
+	alphanumericLowercaseBytes  = "abcdefghijklmnopqrstuvwxyz1234567890"
+	randomSeedVariableName      = "randomSeed"
+	nowVariableName             = "now"
+	ModeEnvironmentVariableName = "AZURE_TEST_MODE"
+)
+
+// Inspired by https://stackoverflow.com/questions/22892120/how-to-generate-a-random-string-of-a-fixed-length-in-go
+const (
+	letterIdxBits = 6                    // 6 bits to represent a letter index
+	letterIdxMask = 1<<letterIdxBits - 1 // All 1-bits, as many as letterIdxBits
+	letterIdxMax  = 63 / letterIdxBits   // # of letter indices fitting in 63 bits
+)
+
+type RecordMode string
+
+const (
+	Record   RecordMode = "record"
+	Playback RecordMode = "playback"
+	Live     RecordMode = "live"
+)
+
+type VariableType string
+
+const (
+	Default             VariableType = "default"
+	Secret_String       VariableType = "secret_string"
+	Secret_Base64String VariableType = "secret_base64String"
+)
+
+// NewRecording initializes a new Recording instance
+func NewRecording(c TestContext, mode RecordMode) (*Recording, error) {
+	// create recorder based on the test name, recordMode, variables, and sanitizers
+	recPath, varPath := getFilePaths(c.Name())
+	rec, err := recorder.NewAsMode(recPath, modeMap[mode], nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the mode is set in the environment, let that override the requested mode
+	// This is to enable support for nightly live test pipelines
+	envMode := getOptionalEnv(ModeEnvironmentVariableName, string(mode))
+	mode = RecordMode(*envMode)
+
+	// initialize the Recording
+	recording := &Recording{
+		SessionName:              recPath,
+		RecordingFile:            recPath + ".yaml",
+		VariablesFile:            varPath,
+		Mode:                     mode,
+		variables:                make(map[string]*string),
+		previousSessionVariables: make(map[string]*string),
+		recorder:                 rec,
+		c:                        c,
+	}
+
+	// Try loading the recording if it already exists to hydrate the variables
+	err = recording.initVariables()
+	if err != nil {
+		return nil, err
+	}
+
+	// set the recorder Matcher
+	recording.Matcher = DefaultMatcher(c)
```

Comment:
So far the test has been able to work perfectly. Appreciate your work @christothes. There are few questions I am curious about 

I see we're using `DefaultMatchers` here. While the default matchers work perfectly in most of the cases, there are few cases where I believe users may need to use custom matchers. So they should be provided options to supply their custom matchers. 

Here are the couple of scenarios which I found while working on integrating `testframework` in `azblob` SDK
1. For upload/download 1MB/2MB/4MB/8MB file to Azure Service (which is a common scenario), the recording file is storing body in request/response which is increasing the size of file and make it infeasible to keep in the repo. And depending on the use-case, I may not be interesting in comparing body content at all.
2. While dealing with Shared Access Signature (SAS, I want to omit the request URL and ignore it from comparisons because those URLs hold secret. I want to provide my custom headers to match apart from 

```go
var ignoredHeaders = map[string]*string{
	"Date":                   nil,
	"X-Ms-Date":              nil,
	"x-ms-date":              nil,
	"x-ms-client-request-id": nil,
	"User-Agent":             nil,
	"Request-Id":             nil,
	"traceparent":            nil,
	"Authorization":          nil,
}
```
supplied in `DefaultMatchers`

```go
matcher := &RequestMatcher{
		context:        testContext,
		IgnoredHeaders: ignoredHeaders,
	}
```

Reply:
Also curious on your opinion on whether we can omit body from the recording as well

Reply:
Some of the most recent commits on this PR added the ability to configure custom matchers.  For example, see the [SetBodyMatcher func here](https://github.com/Azure/azure-sdk-for-go/blob/921c67f7d8afeb91605c4ed912adc462da72636f/sdk/internal/testframework/request_matcher.go#L68). 

For setting custom headers to ignore for matching, you'd add items to the [IgnoredHeaders map](https://github.com/Azure/azure-sdk-for-go/blob/921c67f7d8afeb91605c4ed912adc462da72636f/sdk/internal/testframework/request_matcher.go#L20)

Reply:
As far as eliminating the need to record the body content at all, I think this should be acheivable by setting a [custom BodySanitizer](https://github.com/Azure/azure-sdk-for-go/blob/921c67f7d8afeb91605c4ed912adc462da72636f/sdk/internal/testframework/recording_sanitizer.go#L45) where the "sanitization" is to simply return and empty object (or perhaps one that contains only a single "sanitized" property to make it clear that it was due to sanitization and not a truly empty body)